sklearn/utils/validation.py

```diff
@@ -239,19 +240,30 @@ def _ensure_sparse_format(spmatrix, accept_sparse, dtype, copy,
     spmatrix_converted : scipy sparse matrix.
         Matrix that is ensured to have an allowed type.
     """
-    if accept_sparse in [None, False]:
-        raise TypeError('A sparse matrix was passed, but dense '
-                        'data is required. Use X.toarray() to '
-                        'convert to a dense numpy array.')
     if dtype is None:
         dtype = spmatrix.dtype
 
     changed_format = False
-    if (isinstance(accept_sparse, (list, tuple))
-            and spmatrix.format not in accept_sparse):
-        # create new with correct sparse
-        spmatrix = spmatrix.asformat(accept_sparse[0])
-        changed_format = True
+
+    if isinstance(accept_sparse, str):
+        accept_sparse = [accept_sparse]
+
+    if accept_sparse in [None, False]:
+        raise TypeError('A sparse matrix was passed, but dense '
+                        'data is required. Use X.toarray() to '
+                        'convert to a dense numpy array.')
+    elif (isinstance(accept_sparse, (list, tuple)) and
+          len(accept_sparse) and
```

Comment:
If you pass in an empty list or tuple then it will pass the `isinstance`, but it may be an empty list. `len` ensures it has an element and the next check ensures that the first element is a string.

I can take those checks out however and allow it to fail on its own when it tries to index the list or format with a non-string, if that would be better.